astroplan/constraints.py

```diff
@@ -394,7 +395,79 @@ def compute_constraint(self, times, observer, targets):
                              "MoonSeparationConstraint.")
         return mask
 
+    
+class DarkGreyBrightConstraint(Constraint):
+    """
+    Constrain whether it is dark, gray or bright time or how bright.
+    """
+    # TODO: make a new test
+    @u.quantity_input(horizon=u.deg)
+    def __init__(self, lunar_horizon=90*u.deg, min_moon_illumination=0,
+                                                max_moon_illumination=1):
+        '''
+        Parameters
+        ----------
+        Lunar_horizon : `~astropy.units.Quantity` (optional)
+            The altitude of the ``horizon`` and defaults to 90 degrees       
+        min_moon_illumination : float or `None` (optional)
+            the minimum value of the fraction of the moon illuminated
+        max_moon_illumination : float or `None` (optional)
+            the maximum value of the fraction of the moon illuminated
+        
+        Example
+        -------
+        Constrain the observations to times where the moon :
+
+        >>> from astroplan import Observer
+        >>> from astroplan.constraints import DarkGreyBrightConstraint
+        >>> subaru = Observer.at_site("Subaru")
+        >>> darktime = DarkGreyBrightConstraint(lunar_horizon=-5*u.deg) 
+        >>> greytime = DarkGreyBrightConstraint(max_moon_illumination=.5)
+        '''
+        
+        self.min_illumination = min_moon_illumination
+        self.max_illumination = max_moon_illumination
+        self.horizon = lunar_horizon
+        
+    def _get_lunar_altitudes(self, times, observer, targets):
+        if not hasattr(observer, '_altaz_cache'):
+            observer._altaz_cache = {}
+
+        aakey = (tuple(times.jd), 'sun')
+
+        if aakey not in observer._altaz_cache:
+            # TODO: make sure this works properly
+            # Broadcast the lunar altitudes for the number of targets:
+            altaz = observer.altaz(times, get_moon(times))
+            altitude = altaz.alt
+            altitude.resize(1, len(altitude))
+            altitude = altitude + np.zeros((len(targets), 1))
 
+            observer._altaz_cache[aakey] = dict(times=times,
+                                                altitude=altitude)
+
```

Comment:
I'm using these lines from AtNightConstraint which has them inside a "try:". Do I need to worry about what it is doing?